guides/c.md

```diff
@@ -1,10 +1,98 @@
-The C wrappers should look like the C++ code as much as possible. But, there are going to be differences between the wrappers and the C++ code for many reasons.
+Introduction
+============
 
-- The projects we are wrapping usually have a style guide. Follow that, or make a best guess based on the code already written.
-- Usually the projects use clang-format. It's highly recommended to also use clang-format, and run it whenever the file is saved.
-- In the case of overrides, check if the override is a convenience override or not. If it is not, then ... (TODO: what guideline should be followed?)
+The C wrappers should look like the C++ code as much as possible. But, there are going to be differences between the wrappers and the C++ code for many reasons. When in doubt, refer to to the wrapped library's style guides, or make a best guess based on code already written in either the wrapped library, or other wrappers.
 
-Here are some examples of how to style the C code.
+Formatting
+==========
+
+Use autoformatting where possible. Generally the wrapped libraries use clang-format, so it's best to use that. Also, it is suggested to do formatting on save.
+
+Implementation
+==============
+
+- If there are function overrides, then prefer to wrap the core implementation. If there's multiple implementations, then try to pick a suffix that is clear and understandable.
+
+Error Handling
+==============
+
+TODO
+
+SUGGESTION: Implement something like Rust's `Result<T, E>`, unless the wrapped library already implements error handling that we can inherit from in C.
+
+```c
+enum StatusCode {
+    OK;
+    ERROR;
+};
+typedef struct {
+    StatusCode status;
+    union value {
+        MyType value;
+        Error err;
+    }
+} MyTypeStatus;
+```
+
+Exceptions
+==========
+
+C does not support exceptions, and it's generally considered bad to allow exceptions to pass across the C++ to C boundary (TODO: Why?). So, if wrapped code can cause an exception, then it should be wrapped in code like below.
+
+TODO: This may not catch everything.
+
+```cpp
+ThingStatus mything_do_thing(int thing) {
+    ThingStatus status;
+
+    try {
+        Thing thing = mything::do_thing(int thing);
+
+        status.status = StatusCode.OK;
+        status.value = thing;
+
+        return status
+    } catch (const std::exception& err) {
+        status.status = StatusCode.ERROR;
+        status.err = err;
+        return status;
+    }
+};
+```
+
+Static Asserts
+==============
+
+It is critical to make sure that the types passed through the C++ to C interface are always exactly the same if the type is non-opaque. So, when writing the bindings, there must be a static assert to validate that the C types are the same as the C++ types.
+
+```cpp
```

Comment:
I do think we want to make specific, different recommendations for different kinds of types here (POD, non-trivial etc). I'm curious to see what @luke-titley  is thinking in the other thread about this.

The other thing that needs defining is a pattern (and possibly some utility classes?) for APIs that require subclassing types, e.g. OSL.

Reply:
Sounds good. If we know of any specific cases, then I'll add them in.